app/views/form_drafts/show.haml

```diff
@@ -13,13 +13,16 @@
   = form_tag controller: :forms, action: :submit, id: @draft.form.id do
     = fields_for :user do |u|
       .field
-        .heading Please complete the following personal information.
-        //.explanation You will be able to log in and review your submission.
+        .explanation 
+          This is the personal information we have on file for you.
+          Please fill in any missing or incorrect information.
+          The confirmation will be sent to this email address.
       - [:first_name, :last_name, :email].each do |field|
         .field
           .label= u.label field
-          = u.text_field field, size: 50, required: true,
-            value: @current_user.try(field)
+          = u.text_field field, size: 50,
+            value: @current_user.try(field) || @placeholder.try(field),
```

Comment:
So, to be clear about what I said, I didn't say you shouldn't handle Shibboleth attributes at all in the application controller. As you point out, that's where that needs to be done. My concern was initializing a new user on *every* request. Consider this scenario:

1. Someone gets sent back to us from Shibboleth with attributes. We create a user for them in an instance variable. Great!
2. They reload the page (or do literally anything else). That instance variable is lost. In fact, even if you were to wrap that before action in a conditional so it only executes when we receive attributes from Shibboleth, it wouldn't matter, since instance variables don't persist between requests. And they shouldn't — HTTP is stateless. 

So yes, as you say, you will need to store the shibboleth attributes in a cookie. Then when they view a form, you can auto-fill the fields there.